base/shell.jl

```diff
@@ -88,19 +88,19 @@ function shell_parse(str::AbstractString, interpolate::Bool=true;
                 in_double_quotes = !in_double_quotes
                 i = consume_upto!(arg, s, i, j)
             elseif !in_single_quotes && c == '\\'
-                if !isempty(st) && peek(st)[2] in ('\n', '\r')
+                if !isempty(st) && (peek(st)::Pair{Int,Char})[2] in ('\n', '\r')
```

Comment:
Hmm, I am a bit surprised these type annotations are needed. `st` should have a concrete type, right? So is `peek` itself not type stable?

Reply:
yeah, `peek(s::Stateful)` may return `nothing` when `s` is empty.

Reply:
But since `nothing[2]` is an error, this would also be fixed via https://github.com/JuliaLang/julia/pull/40880?

Reply:
`peek` returns `nothing` if the iterator has been exhausted. So one could replace the `!isempty` check by branching on the return type instead and that should allow to get away without the type-assertion, but I'm not sure that's better.

Reply:
> But since nothing[2] is an error, this would also be fixed via #40880?

No, #40880 only works _after_ a target local variable is used:
```julia
                if !isempty(st) && (p = peek(st); p[2] in ('\n', '\r'))
                    p # p::Pair{Int,Char} after #40880
```

Unfortunately, it's invalid if the compiler automatically optimizes `peek(st)[2] in ('\n', '\r')` into something like `peek(st)::Pair{Int,Char}[2] in ('\n', '\r')`, since it changes the semantics (the possible exception is changed to `TypeError` from `MethodError`).

> `peek` returns `nothing` if the iterator has been exhausted. So one could replace the `!isempty` check by branching on the return type instead and that should allow to get away without the type-assertion, but I'm not sure that's better.

Yeah, I'd like to go with type assertion for now. I think maybe we can refine our iteration protocol so that our type lattice can recognize the remaining length of an iterator ? It'd look very similar to array shape analysis.